File: scala-3.6/tax-system/src/main/scala/com/rbleggi/taxsystem/service/TaxCalculator.scala
package com.rbleggi.taxsystem.service

import com.rbleggi.taxsystem.model.{Product, State, TaxRule}
import com.rbleggi.taxsystem.specification.{ProductSpecification, StateSpecification, YearSpecification}

class TaxCalculator(rules: List[TaxRule]):

  def calculateTotalPrice(product: Product, state: State, year: Int): Double =
    val productSpec = ProductSpecification(product)
    val stateSpec = StateSpecification(state)
    val yearSpec = YearSpecification(year)

    val combinedSpec = productSpec.and(stateSpec).and(yearSpec)

    rules
      .filter(combinedSpec.matches)
      .map(_.calculateTax())
      .headOption
      .map(tax => product.price + tax)
      .getOrElse(product.price)

File: kotlin/tax-system/src/main/kotlin/com/rbleggi/taxsystem/calculator/TaxCalculator.kt
package com.rbleggi.taxsystem.calculator

import com.rbleggi.taxsystem.model.TaxRule

class TaxCalculator(private val rules: List<TaxRule>) {
    fun tax(product: String, price: Double, state: String): Double =
        rules.firstOrNull { it.product == product && it.state == state }?.tax(price) ?: 0.0

    fun totalPrice(product: String, price: Double, state: String): Double =
        price + tax(product, price, state)
}
